### ISSUE TITLE
Make MatomoTracker a non final class

### ISSUE THREAD

**Issue by lukas2** [piwik/piwik-sdk-ios#287]
It would be great if we could subclass MatomoTracker to be able to mock it for unit tests. I understand the final-keyword gives us slightly better performance, but I would say it's fair to trade it for the ability to ensure tracking is happening correctly with unit tests.

(We can use a wrapper or mock using the protocol oriented approach but it's a bit cumbersome.)

**Comment by brototyp** [comment]
Hi @lukas2, thanks a lot for bringing this up. That is a very interesting use case I haven't thought about yet. Before going this route let's check what other options there are to get to the same result, since, aside from slight performance loss, there are downsides to making the `MatomoTracker` open.

Have you thought about using a custom queue? Especially for just testing, a custom queue might be a super easy way to get every single event being generated by the Framework with just implementing a single function (`public func enqueue(events: [Event], completion: (()->())?)`) and mocking the rest. Custom queues were enabled by [this](https://github.com/matomo-org/matomo-sdk-ios/pull/279/files) PR. Unfortunately they aren't really documented, but there is an example implementation that might get you up to speed over [here](https://github.com/matomo-org/matomo-sdk-ios/pull/279/files#diff-80da05c152f246838a5717cbcaad9941R4).

I am looking forward to your feedback on this.

**Comment by lukas2** [comment]
Hi, thanks for the quick and helpful answer. I was able to test my implementation using a custom queue. It was a painless and straightforward process. (My learning here: Before opening a ticket, take a quick look at the interface of dependencies and check if there's another way to get what I want.)

Still, philosophically I guess, I'd prefer (almost) no class to be final. But that is probably my first love speaking: Ruby. ;)

**Comment by brototyp** [comment]
Hi @lukas2, awesome! Super glad to help and that we actually had a solution for your needs in place. I think we should definitely document the custom queues as well as possibly write a short paragraph in the FAQ in regards of testing the MatomoTracker.

I am open to arguments towards changing the class to open but I would hesitate to open it on principle. Especially with open source, there is a responsibility to kind of support what you made open, so the more we open up the MatomoTracker, the more we would need to support and bring along. 

I think we can keep this issue as a kind of documentation about our thoughts towards this question.

**Comment by lukas2** [comment]
I won't argue for opening the class now, because the queue-way works for me, but I'm curious how opening the class would mean you have to support it more. I'd argue it's the other way round, it would make it easier for devs like me to help themselves. I'm just trying to use this as an opportunity to learn about why declaring things `final` can be of benefit.

**Comment by brototyp** [comment]
Hey @lukas2, sorry for not replying for a little longer. 

We are trying to write the MatomoTracker as good as possible, have easy and clean code with as few as possible side effects. But there will be invariants within the SDK that we aren't aware of. The more open we make the MatomoTracker class or the SDK in general, the higher the risk, that users of it are relying on any of these. If we don't want to break their usage pattern we would that first be aware of these invariants and then make sure we continue to support the usage patterns. For an open class all our dependencies and assumptions are much more implicit then for a clear API that we wrote our selves.

I am not sure if I was able to bring my point across here. The risks are much more unknown than in other situations.
